examples/tutorial/TUTORIAL.md

```diff
@@ -0,0 +1,191 @@
+## Introduction to gRPC-Haskell
+
+*This tutorial assumes that you already have a basic understanding of gRPC as well as Haskell.*
+
+This will go through a basic example of using the library, with the `arithmetic` example in the `examples/arithmetic` directory. After cloning this repository, it would be a good idea to run `stack haddock` from within the repository directory to generate the documentation so you can read more about the functions and types we're using as we go. Also remember that [typed holes](https://wiki.haskell.org/GHC/Typed_holes) can be very handy.
+
+The gRPC service we will be implementing provides two amazing functions:
+
+1. `Add`, which adds two integers.
+2. `RunningSum`, which receives a stream of integers from the client and finally returns a single integer that is the sum of all the integers it has received.
+
+You can run the examples by running `stack exec arithmetic-server` and `stack exec arithmetic-client`.
+
+### Library Organization
+
+**tl;dr: you probably only need to import `Network.GRPC.HighLevel.Generated`.** Other modules are exposed for advanced users only.
+
+This library exposes quite a few modules, but you won't need to worry about most of them. They are currently organized based on the level of abstraction they afford over using the C [gRPC Core library](http://www.grpc.io/grpc/core/) directly:
+
+* *`Unsafe`* modules directly wrap functions in the gRPC Core library. Using them directly is like using C: you need to think about memory management, pointers, and so on. The rest of the library is built on top of these functions and users of gRPC-haskell should never need to deal with the `Unsafe` modules directly.
+* *`LowLevel`* modules still require an understanding of the gRPC Core library, but guarantee memory and thread safety. Only advanced users with special requirements would use `LowLevel` modules directly.
+* *`HighLevel`* modules give you an opinionated Haskell interface to gRPC that should cover most use cases while (hopefully) being easy to use. You should only need to import the `Network.GRPC.HighLevel.Generated` module to start using the library. If you need to import other modules, we probably forgot to re-export something and you should open an issue or PR.
+
+### Getting started
+
+To start out, we need to generate code for our protocol buffers and RPCs.
+
+```
+stack exec -- compile-proto-file --proto examples/echo/echo.proto > examples/echo/echo-hs/Echo.hs
```

Comment:
How do users get `compile-proto-file` (does it come from `proto3-suite` as a dependency?